### ISSUE TITLE
如何在完全离线的情况下加载msdataset？

### ISSUE THREAD

**Issue by Lotayou** [modelscope/modelscope#112]
背景：实验室研发，本地电脑联网但没有显卡，服务器有卡但不通外网。因此需要先从本地预先下载Modelscope的数据集，迁移到服务器中，再直接从服务器离线读取本地缓存。

以官方OFA-OCR finetune教程为例，首先本地下载了ocr-fudan数据集，代码如下：

```
train_dataset=MsDataset.load(
        'ocr_fudanvi_zh',
        namespace='modelscope',
        split='train'),
    eval_dataset=MsDataset.load(
        'ocr_fudanvi_zh',
        namespace='modelscope',
        split='validation')
```

下载成功后，将整个/root/.cache/modelscope/hub/datasets挪到实验室的内网服务器中，本地的数据结构如下
![企业微信截图_1676550998371](https://user-images.githubusercontent.com/33449901/219366694-23456358-dd3b-40c0-9af1-25d4c8b75e54.png)

之后运行同样的数据加载代码会提示url无法连接。尝试直接指定数据集的缓存路径，代码如下
```
test_dataset = MsDataset.load(
            '/root/.cache/modelscope/hub/datasets/ocr_fudanvi_zh',
            namespace="modelscope",
            download_mode=DownloadMode.REUSE_DATASET_IF_EXISTS,
            split='test')
```
依然无法绕过外网连接检测。
![企业微信截图_1676551297816](https://user-images.githubusercontent.com/33449901/219367686-69d8aca1-0950-489e-852e-8363fa754313.png)

求问应该如何改造代码才能在内网环境下直接读取本地预先下好的数据集，跳过联网检查的流程？感谢~

**Comment by wangxingjun778** [comment]
如果是针对“已经下载了数据集到本地缓存，并且希望在无网络环境下再次加载”这个场景，目前modelscope的数据集模块暂时不支持如此使用，因为sdk端默认需要保持与hub端的元数据和数据文件同步，以及相关鉴权信息的获取等。另外，对于该数据集，由于是托管到外部host，所以在使用MsDataset.load()加载后得到的对象，在做iter时，每访问一条数据，都需要从远端服务器download图像文件，所以亦会限制在上述场景下使用。

如果确实需要在无网络环境下使用，可以参考以下脚本，来间接加载缓存数据：

Step1：遍历数据集的各个split，并重写到本地对应路径：
`  import os
    from modelscope.msdatasets import MsDataset

    work_dir = '/to/path/your-data-path'
    ds = MsDataset.load("ocr_fudanvi_zh", subset_name='scene', namespace="modelscope")

    def process(dataset, split_name, work_dir):
        ds_split = dataset[split_name]

        split_img_dir = os.path.join(work_dir, split_name, 'images')
        os.makedirs(split_img_dir, exist_ok=True)

        idx = 1
        data_list = ['image_id,label,image']
        for item in ds_split:
            if idx > 100:  # todo
                break
            print(f'>{split_name}-idx:', idx)
            img = item.get('image')
            img_path = os.path.join(split_img_dir, f"{item.get('image_id')}.png")
            img_path_relative = os.path.join(split_name, 'images', f"{item.get('image_id')}.png")
            if os.path.exists(img_path):
                idx += 1
                continue
            img.save(img_path)
            data_list.append(item.get('image_id') + ',' + item.get('label') + ',' + img_path_relative)
            idx += 1
        with open(os.path.join(work_dir, f'{split_name}.csv'), 'w') as f:
            for line in data_list:
                f.write(line + '\n')

    splits = ['train', 'test', 'validation']
    for split in splits:
        process(ds, split, work_dir)
`

Step2:  将上述work_dir中的所有数据搬到目标服务器上去，并使用load函数加载：
`  work_dir = '/to/path/your-data-path'
    ds_test = MsDataset.load('csv', data_files=[os.path.join(work_dir, 'test.csv')])
    print(next(iter(ds_test)))
`

**Comment by Lotayou** [comment]
感谢及时回答，问题解决了。
顺便想探讨下，为什么会有这样的设计，先下载一半元信息然后做iter时再从远端服务器上拉图片呢？这样引入了很多网络下载开销，感觉会拖慢训练效率

**Comment by wangxingjun778** [comment]
> 感谢及时回答，问题解决了。 顺便想探讨下，为什么会有这样的设计，先下载一半元信息然后做iter时再从远端服务器上拉图片呢？这样引入了很多网络下载开销，感觉会拖慢训练效率

--这里主要是由数据集特点决定的，该数据集的贡献者选择了“托管到公开host”（所谓公开host，即贡献者选择不托管在modelscope的存储资源中，而是引用了外部资源，如aliyun oss/aws s3/huggingface等），那么MsDataset.load()加载数据集时会follow其元数据中的py script文件的加载逻辑。
参考ocr_fudanvi_zh数据集的py script： https://modelscope.cn/datasets/modelscope/ocr_fudanvi_zh/file/view/master/ocr_fudanvi_zh.py ， 
在_split_generators()中，作者使用了dl_manager.iter_files()来对托管在oss上的zip文件做iter，然后在_generate_examples()中对得到的file做open操作，所以会出现我们前文讨论过的情况。

同时，更多的情况是“托管到modelscope”类型的数据集，这种通常是可以直接全量加载到本地缓存，或streaming load的方式来使用，故不会出现上述问题。

以上，供参考 ~~
